src/core/truthTable/truth_table.cpp

```diff
@@ -0,0 +1,252 @@
+            if (result.second == false)
+                result.first->second++;
+        }
+
+        for (auto& elem: outputFreq) {
+            outputFreq[elem.first] = ceil(log2(elem.second));
+        }
+
+        // if the given truth table is not reversible
+
+        if (outputFreq.size() != inOutCube.size()) {
+            std::vector<unsigned> maxEncSize;
+
+            unsigned encSize;
+
+            truthTable::value_type emptyVal;
+
+            truthTable::cube_type emptyVec;
+
+            truthTable::cube_type codedVec;
+
+            struct syrec::MinHeapNode *left, *right, *top;
+
+            std::priority_queue<syrec::MinHeapNode*, std::vector<syrec::MinHeapNode*>, syrec::compare> minHeap;
+
+            for (auto& elem: outputFreq)
+                minHeap.push(new syrec::MinHeapNode(elem.first, elem.second));
+
+            while (minHeap.size() != 1) {
+                left = minHeap.top();
+                minHeap.pop();
+
+                right = minHeap.top();
+                minHeap.pop();
+
+                top = new syrec::MinHeapNode(emptyVec, (std::max(left->freq, right->freq) + 1));
+
+                top->left  = left;
+                top->right = right;
+
+                minHeap.push(top);
+            }
+
+            //encode the output if necessary
+
+            hufCodes(minHeap.top(), codedVec, encInOut);
+
+            //increase the lengths of the encoded outputs with dont cares if required
+
+            for (auto& i: encInOut) {
+                maxEncSize.push_back(i.second.size());
+            }
+
+            encSize = *(std::max_element(maxEncSize.begin(), maxEncSize.end()));
+
+            for (auto& i: encInOut) {
+                if (encSize > i.second.size()) {
+                    for (unsigned j = 0; j <= (encSize - i.second.size()); j++) {
+                        i.second.push_back(emptyVal);
+                    }
+                }
+            }
+
+            //replace the outputs with the corrsponding encoded outputs
+
+            for (auto& i: inOutCube) {
+                inOutCube[i.first] = encInOut[i.second];
+            }
+        }
+
+        //Append zeros to the input if the output length (either encoded or not) is higher compared to the input
+
+        for (auto& i: inOutCube) {
+            if (i.second.size() > i.first.size()) {
+                for (unsigned j = 0; j <= (i.second.size() > i.first.size()); j++) {
+                    truthTable::cube_type dummy = i.first;
+
+                    auto node = inOutCube.extract(i.first);
+
+                    node.key() = insert_zero(dummy);
+
+                    inOutCube.insert(std::move(node));
+                }
+            }
+        }
+
+        tt.clear();
+
+        //update the truth table
+
+        for (auto& i: inOutCube) {
+            truthTable::cube_type firstVal = i.first;
+
+            truthTable::cube_type secondVal = i.second;
+
+            tt.add_entry(firstVal, secondVal);
+        }
+    }
```

Comment:
this whole function is a mess with linter warnings sprinkled all over it. no chance to review it like that.